Implement Stopwatch.ToString()

ericmutta opened the pull request:
Override the `ToString()` method for the `Stopwatch` class to return a reasonable default representation of the elapsed time.

Fix #87449

Commit:
Implement Stopwatch.ToString()

For the tests, I added a few one-liners to the existing tests
where the Stopwatch.Elapsed property was easy to check.
I also added the OverridesToString() test to check for cases
where the ToString() method is called implicitly. This confirms
that people can write "Console.WriteLine(sw)" where "sw" is
an instance of the Stopwatch class, and get the elapsed time
printed.

Fix #87449

tannergooding commented on src/libraries/System.Runtime.Extensions/tests/System/Diagnostics/Stopwatch.cs:
```diff
@@ -96,6 +99,25 @@ public static void StartNewAndRestart()
+        [Fact]
+        public static void OverridesToString()
+        {
+            // In this test we use string interpolation witch a Stopwatch instance to trigger
```
```suggestion
            // In this test we use string interpolation which a Stopwatch instance to trigger
```

ericmutta replied:
@tannergooding thanks for catching this, it is actually a typo on my part, I wanted to write:

> ...we use string interpolation **WITH** a Stopwatch instance...

I've pushed another commit with the typo fix 👍

saul commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs:
```diff
@@ -89,6 +89,8 @@ public void Restart()
             _isRunning = true;
         }
 
+        public override string ToString() => Elapsed.ToString();
```
Wonder if it's worth exposing `IsRunning` here as well?

ericmutta committed:
Merge branch 'dotnet:main' into issue-87449

ericmutta committed:
Fix typo in Stopwatch.cs

Fix typo in the comment within the OverridesToString() test.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs:
```diff
@@ -89,6 +89,8 @@ public void Restart()
             _isRunning = true;
         }
 
+        public override string ToString() => Elapsed.ToString();
```
Since Stopwatch isn't sealed, technically this should also be added to the ref:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/ref/System.Runtime.cs
Otherwise, a type that derives from it and calls base.ToString() might end up skipping this override.  And because it's publicly visible, it should also have XML comments added to it to seed the docs

ericmutta replied:
Thanks @stephentoub for the guidance. I have added the XML comments and updated the `System.Runtime.cs` file as well (for now I just copied what was being done with other classes that have a `ToString()` override, but later on, I plan to look into the whole `ref` mechanism to understand how it works).

Commit:
Add XML docs for the ToString() method.

The <returns> tag is somewhat redundant since the <summary> tag
adequately describes what's going on here, but since the convention
is to have both tags, I have used the <returns> tag to clarify that
the returned string renders the elapsed time using the same format as
TimeSpan.ToString().

Commit:
Add ToString() override for Stopwatch class.

See https://github.com/dotnet/runtime/pull/87812#discussion_r1235940302

stephentoub commented:
Thanks.

danmoseley commented:
@ericmutta Thanks for this. Given this is your first contribution it would be really helpful to now tune our docs wherever you were led astray or got stuck. Is there anything like that you could offer in a PR? Anything from typos to a big change, as you suggest.

ericmutta commented:
I remember about 23 years ago, I was in college and I almost got hit by those [red double-decker buses](https://en.wikipedia.org/wiki/AEC_Routemaster) they have in England. I wasn't paying attention to traffic when crossing the road because I was engrossed in an article about this new thing called C#, back when people were still debating whether it's pronounced see-sharp, see-hash or see-pound.

I immediately fell in love with the .NET Framework and years later it empowered me, a young African software engineer without a dollar to his name, [to score a couple hundred thousand dollars in prize money](https://www.thecitizen.co.tz/tanzania/news/young-man-in-sh24-trillion-dream-2494076) 🔥  So, as you can imagine, after making [over 70 minor contributions to the dotnet repos](https://github.com/pulls?q=is%3Apr+author%3Aericmutta+archived%3Afalse+is%3Aclosed+org%3Adotnet+), it is both a pleasure and **a great privilege to finally contribute some actual code** 💪

Thanks @danmoseley for giving me the chance to work on this, and to the PR reviewers (@stephentoub, @tannergooding & @saul) for making sure everything was in order before merging 🙏

### Using Codespaces To Contribute

@danmoseley the next contribution I would like to try is improving the [Codespaces experience](https://github.com/dotnet/runtime/blob/main/docs/workflow/Codespaces.md). I was really drawn to the idea of using a temporary VM when making a simple contribution like `Stopwatch.ToString()` but that didn't work out too well:

1. **I got build errors out of the box**. This may be due to something I overlooked but I would like to try again to make sure that anyone following the workflow docs you pointed me to, **can do a clean build without any custom tweaks.**

2. **Using the JetBrains Rider IDE**. Since SQL Server refuses to install on those cheap $5 VMs you can get from some cloud providers these days, I switched to using Postgres as my backend database. And since Visual Studio doesn't offer any plugins or support for Postgres, I switched to using JetBrains Rider as my IDE for both C# and Postgres work. Finally, since .NET 7 doesn't support Windows 8.1, I had to free up some space in order to upgrade to Windows 10 and in the process **I removed Visual Studio completely from my machine**. The [Codespaces docs](https://docs.github.com/en/codespaces/developing-in-codespaces/using-github-codespaces-in-your-jetbrains-ide) say it is possible to use JetBrains IDEs for remote development, **but it needs a codespace running an SSH server** so I would like to take a stab at [updating the devcontainer.json file](https://github.com/dotnet/runtime/blob/main/docs/workflow/Codespaces.md#updating-dotnetruntimes-codespaces-configuration) so [it installs SSH](https://docs.github.com/en/codespaces/developing-in-codespaces/using-github-codespaces-with-github-cli#ssh-into-a-codespace), then see if I can use JetBrains Rider to get a full IDE experience locally while all the runtime code and build work happens remotely. I am acutely aware that a competing product is involved in all this, so I will understand if I don't get much help here 😄

Oh, there's also a broken image link in the [Create a Codespace](https://github.com/dotnet/runtime/blob/main/docs/workflow/Codespaces.md#create-a-codespace) section. I wanted to fix it but I wasn't sure what was being pointed to, so maybe my namesake Eric (@eerhardt) can assist (Git blame shows [he touched that link](https://github.com/dotnet/runtime/blame/main/docs/workflow/Codespaces.md#L15) 9 months ago) 👍 

### Commit Message Formatting

The [contribution docs](https://github.com/dotnet/runtime/blob/main/CONTRIBUTING.md#commit-messages) offer guidance on how to format commit messages. I used Visual Studio to implement the `Stopwatch.ToString()` change and when using the integrated Git experience to write the commit messages, I noticed that **the commit message box has no hints about what line/column number the caret is on**. So to make sure the first commit message line was no longer than 50 characters and that other lines were no longer than 72 characters, **I actually had to type the commit message in an external editor such as Notepad++ then paste it back into Visual Studio!**

Maybe you guys can internally nag/ask the team responsible to implement the `.editorconfig` equivalent for the commit message box so projects like this one can specify the rules to be followed and have them enforced automatically by the IDE 👍

### Where can I find the code and tests for `System.Foo.Bar`?

By far the biggest challenge as a first time contributor is just finding stuff. To this moment, I am still not sure which project/solution file to open in Visual Studio in order to get a full editing and debugging experience for the [Stopwatch.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs) file.

@danmoseley if it wasn't for the links you shared with me, I wouldn't have known where to begin! Maybe one way to fix this centrally, is to have each area owner add links to the `src` and `tests` folders for their respective area. These links could be added to the Notes column of the [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) file. **The idea here is that by looking at ONE LOCATION, you can know who to tag and where to find the solution/project files for both the source code and the tests for that particular area.**

### Conclusion

**Overall this has been a great experience for me** and I would certainly encourage others to take that leap and contribute some code whenever the chance arises. It takes a little patience (and _a lot_ of disk space 😄) but what could be cooler than knowing every machine that ever installs .NET 8+ is going to run code that you wrote? That right there, is **priceless** 🚀

danmoseley commented:
@ericmutta thank you for the appreciation, it's great to hear your story and great to have new contributors.

1. We'd welcome improvements to the Codespaces experience for folks whatever the tools they choose. Contributors here use all kinds of tools (and OS). The broken link must have been to the create codespaces button. We could commit that as a small image instead.
2. For the commit message formatting, can you open a VS feedback issue? I wasn't aware there was editorconfig support to define commit message width - sounds like a good idea to wire it up.
3. Great suggestion, sharing with @ericstj @jeffhandley @ViktorHofer.
